design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,321 @@
+  cluster.
+
+## Background
+
+Currently, Crossplane models portable cloud provider managed services as
+*managed resources* that can be either statically or dynamically provisioned
+using resource *classes* and *claims*. One such resource type is a Kubernetes
+cluster, which uses a `KubernetesCluster` claim and cloud provider-specific
+managed resource types (e.g. `GKECluster`, `AKSCluster`, `EKSCluster`, etc.). It
+is appropriate for Kubernetes clusters to be modeled in the same manner as other
+managed services due to the fact that they are presented by the provider in the
+same fashion. However, in Crossplane, Kubernetes clusters are unique to other
+managed service offerings in that they can also be used as a target for running
+workloads.
+
+Workloads are provisioned using the `KubernetesApplication` type. It is
+namespace-scoped and is scheduled to a target cluster using labels to match a
+`KubernetesCluster` claim in its namespace. Because managed resources may only
+be bound to a single claim, the claim is namespaced, and `KubernetesApplication`
+resources can only be scheduled to `KubernetesCluster` claims in their
+namespace, any Kubernetes managed resource offering may only be consumed (i.e.
+deployed to) from a single namespace.
+
+In order to fully conceptualize the usage of Kubernetes clusters in Crossplane,
+it is helpful to separate the notions of a **Kubernetes cluster as a managed
+service** and a **Kubernetes cluster as a target for deploying workloads**. With
+this distinction made, we can assert that this proposal in no way modifies the
+managed service notion of a Kubernetes cluster, but instead formalizes this
+concept of a Kubernetes cluster as a deployment target.
+
+## Goals
+
+Because it is unlikely that Crossplane users will always want a 1-to-1
+namespace-to-cluster consumption model, it is desirable to support deploying
+workloads in any namespace to any Kubernetes cluster. However, it is also
+desirable to continue to treat managed Kubernetes services as a managed resource
+that is supplied by cloud providers. For this reason, any solution should be
+additive to the existing model.
+
+As a secondary goal for this design, it should be possible to use existing
+Kubernetes clusters (wherever they may be running) that were not provisioned by
+Crossplane as targets for deploying workloads. This concept has frequently been
+referred to as a "Bring Your Own Cluster" scenario.
+
+## Proposal
+
+This proposal can be separated into three main categories:
+1. The implementation of a new `KubernetesTarget` custom resource and
+   corresponding controllers.
+1. The modification of the scheduling behavior for `KubernetesApplication`.
+1. The implementation of a controller that automatically creates a
+   `KubernetesTarget` resource when a `KubernetesClaim` is bound to a Kubernetes
+   cluster managed resource. 
+
+### KubernetesTarget
+
+The purpose of the `KubernetesTarget` resource is to effectively "publish"
```

Comment:
With namespaced claims now exclusively matching to cluster-scoped resource classes, the publishing of `KubernetesTarget` to a namespace seems more aligned with the previous portable resource classes approach where `KubernetesApplication` is matched to a namespaced `KubernetesTarget` much like claims were matched to namespaced portable resource classes.

Would it make sense for `KubernetesTarget` be cluster-scoped to be more consistent with the claims/classes model we have today? That would impact the ability for app teams to create their own clusters as cattle, but if app teams are creating their own Kubernetes clusters you could argue they should have their own Crossplane instance as well - esp. given they have joint access to all the cluster-scoped resource classes which references the provider secret.

Reply:
If we made `KubernetesTarget` cluster scoped we would no longer support (or would at least make quite awkward) the concept of a namespace owner provisioning their own clusters to deploy their workloads to. I think that would be a shame - I'd prefer to leave that avenue open while also allowing infrastructure operators to supply a set of clusters for folks to use. We'd also be preventing infrastructure operators from having any control over who gets to use which cluster. If a target exists, it's fair game for any namespace to deploy applications to.

If we do decide `KubernetesTarget` should be cluster scoped, then `KubernetesTarget` should not  exist. A cluster scoped `KubernetesTarget` exists, and is named `Provider`.

Reply:
Some carry-on effects of making `KubernetesTarget` cluster scoped would be:

* Namespace owners cannot BYO cluster; only infrastructure owners can do this.
* There's close to no reason for the `KubernetesCluster` claim to exist anymore. There would be little reason for namespace owners to claim Kubernetes clusters, because they could not use them as deployment targets.